src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2020 Aaron Anderson, Jalex Stark. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson, Jalex Stark.
+-/
+
+import linear_algebra.char_poly
+import linear_algebra.matrix
+import ring_theory.polynomial.basic
+import algebra.polynomial.big_operators
+
+/-!
+# Characteristic polynomials
+
+We give methods for computing coefficients of the characteristic polynomial.
+
+## Main definitions
+
+- `degree_char_poly_eq_dim`
+- `det_from_char_poly`
+- `trace_from_char_poly`
+
+## Implementation details
+
+-/
+
+noncomputable theory
+
+universes u v w z
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type v} [fintype n] [decidable_eq n]
+variables {α : Type v} [decidable_eq α]
+
+open finset
+open polynomial
+
+section fixed_points
+-- TODO: move this section to be with other combinatorial facts about the symmetric group
+lemma gt_one_nonfixed_point_of_nonrefl {σ : equiv.perm n} (h : σ ≠ equiv.refl n) :
+1 < (filter (λ (x : n), ¬ σ x = x) univ).card :=
+begin
+  rw one_lt_card_iff,
+  contrapose! h, ext, dsimp,
+  have := h (σ x) x, simp only [true_and, mem_filter, equiv.apply_eq_iff_eq, mem_univ, ne.def] at this,
+  tauto,
+end
+
+lemma lt_card_sub_one_fixed_point_of_nonrefl {σ : equiv.perm n} (h : σ ≠ equiv.refl n) :
+(filter (λ (x : n), σ x = x) univ).card < fintype.card n - 1:=
+begin
+  have hun := @filter_union_filter_neg_eq _ (λ (x : n), σ x = x) _ _ _ univ,
+  have hin : (filter (λ (x : n), σ x = x) univ) ∩ (filter (λ (x : n), ¬ σ x = x) univ) = ∅
+    := filter_inter_filter_neg_eq univ,
+  rw ← disjoint_iff_inter_eq_empty at hin,
+  rw fintype.card, conv_rhs { rw ← hun },
+  rw card_disjoint_union hin,
+  have := gt_one_nonfixed_point_of_nonrefl h, omega,
+end
+
+end fixed_points
+
+variable {M : matrix n n R}
+
+lemma nat_degree_char_matrix_val [nontrivial R] (i j : n) :
+  (char_matrix M i j).nat_degree = ite (i = j) 1 0 :=
```

Comment:
Perhaps reverse to `char_matrix_val_nat_degree`? (We have some flexibility here, but I think the consensus is that when using dot notation it's okay to name `X.Y` as `X_Y` rather than `Y_X`, even though older stuff in mathlib that predates dot notation doesn't do so.)

We also talked about whether matrix entries should be `val` or `apply` (which I prefer as more consistent with the rest of the library).

Reply:
Also -- it's reasonable to use the name-ordering flexibility to make lemma names more locatable -- it's valuable to put the "subject" of the theorem first where it works, which here is clearly `char_matrix` and not `nat_degree`.